### PR TITLE
[MachineOutliner] Sort by Benefit to Cost Ratio

### PR DESCRIPTION
**Pull request by xuanzhang816**
This PR depends on https://github.com/llvm/llvm-project/pull/90260

We changed the order in which functions are outlined in Machine Outliner. 

The formula for priority is found via a black-box Bayesian optimization toolbox. Using this formula for sorting consistently reduces the uncompressed size of large real-world mobile apps. We also ran a few benchmarks using LLVM test suites, and showed that sorting by priority consistently reduces the text segment size.

|run (CTMark/)   |baseline (1)|priority (2)|diff (1 -> 2)|
|----------------|------------|------------|-------------|
|lencod          |349624      |349264      |-0.1030%     |
|SPASS           |219672      |219480      |-0.0874%     |
|kc              |271956      |251200      |-7.6321%     |
|sqlite3         |223920      |223708      |-0.0947%     |
|7zip-benchmark  |405364      |402624      |-0.6759%     |
|bullet          |139820      |139500      |-0.2289%     |
|consumer-typeset|295684      |290196      |-1.8560%     |
|pairlocalalign  |72236       |72092       |-0.1993%     |
|tramp3d-v4      |189572      |189292      |-0.1477%     |

This is part of an enhanced version of machine outliner -- see [RFC](https://discourse.llvm.org/t/rfc-enhanced-machine-outliner-part-1-fulllto-part-2-thinlto-nolto-to-come/78732).

### CODE REVIEW

File: llvm/test/CodeGen/AArch64/machine-outliner-sort-per-priority.ll [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+;   ```
+;     mov     w6, #6
+;     mov     w7, #7
+;     b
+;   ```
+; It has:
+;   - `SequenceSize=12` and `OccurrenceCount=4`
+;   - each Candidate has `CallOverhead=4` and `FrameOverhead=0`
+;   - `NotOutlinedCost=12*4=48` and `OutliningCost=4*4+12+0=28`
+;   - `Benefit=120-96=20` and `Priority=48/28=1.71`
```
Comment (kyulee-com) [inline llvm/test/CodeGen/AArch64/machine-outliner-sort-per-priority.ll:28]: Should the benefit computation be `Benefit=48-28=20`?

**Review by kyulee-com — COMMENTED** [review]
Can you please add a MIR test to clearly demonstrate the actual instruction sequences? This would help ensure the implementation aligns with our expectations with regard to priority computation.

File: llvm/test/CodeGen/AArch64/machine-outliner-sort-per-priority.ll (hunk 1 above)
Comment (xuanzh-meta) [reply llvm/test/CodeGen/AArch64/machine-outliner-sort-per-priority.ll:28]: Yes! Thanks for catching this!

File: llvm/test/CodeGen/AArch64/machine-outliner-sort-per-priority.mir [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    TCRETURNdi @foo, 0, csr_darwin_aarch64_aapcs, implicit $sp, implicit killed $w0, implicit killed $w1, implicit killed $w2, implicit killed $w3, implicit killed $w4, implicit killed $w5, implicit killed $w6, implicit killed $w7
+
+...
+---
+# CHECK-LABEL: name:            f4
+# CHECK-LABEL: bb.0:
+# CHECK-NEXT: liveins: $lr
+# CHECK-NEXT: {{ $}}
+# CHECK-NEXT: $x5 = ORRXrs $xzr, $lr, 0
+# CHECK-NEXT: BL @OUTLINED_FUNCTION_1, implicit-def $lr, implicit $sp, implicit-def $lr, implicit-def $w0, implicit-def $w1, implicit-def $w2, implicit-def $w3, implicit-def $w4, implicit $sp, implicit $xzr, implicit $x5
```
Comment (kyulee-com) [inline llvm/test/CodeGen/AArch64/machine-outliner-sort-per-priority.mir:110]: I think for the better readability, you can simplify it while dropping all implicit operands:
```
# CHECK-NEXT: BL @OUTLINED_FUNCTION_1
```

You can also do the similar thing for the following case:
```
# CHECK-NEXT: TCRETURNdi @OUTLINED_FUNCTION_0
```

File: llvm/test/CodeGen/AArch64/machine-outliner-sort-per-priority.mir [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+...
+---
+# CHECK-LABEL: name:            f5
+# CHECK-LABEL: bb.0:
+# CHECK-NEXT: liveins: $lr
+# CHECK-NEXT: {{ $}}
+# CHECK-NEXT: $x5 = ORRXrs $xzr, $lr, 0
+# CHECK-NEXT: BL @OUTLINED_FUNCTION_1, implicit-def $lr, implicit $sp, implicit-def $lr, implicit-def $w0, implicit-def $w1, implicit-def $w2, implicit-def $w3, implicit-def $w4, implicit $sp, implicit $xzr, implicit $x5
+# CHECK-NEXT: $lr = ORRXrs $xzr, $x5, 0
+# CHECK-NO: @OUTLINED_FUNCTION_0
```
Comment (kyulee-com) [inline llvm/test/CodeGen/AArch64/machine-outliner-sort-per-priority.mir:141]: I'm not sure if this is the right match. `CHECK-NO` doesn't apply. If you're meant to not match, use `CHECK-NOT`.

Comment (xuanzh-meta) [reply llvm/test/CodeGen/AArch64/machine-outliner-sort-per-priority.mir:141]: Oops. Fixed.

**Review by kyulee-com — APPROVED** [review]
lgtm, but please wait a bit for others' comments if any.

**Pull request merged**